docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -38,14 +38,14 @@ The storage model presented here doesn't deal with data structure nor serializat
 
 Separation of storage and commitment (by the SMT) will allow the optimization of different components according to their usage and access patterns.
 
-`SS` (SMT) is used to commit to a data and compute merkle proofs. `SC` is used to directly access data. To avoid collisions, both `SS` and `SC` will use a separate storage namespace (they could use the same database underneath). `SC` will store each `(key, value)` pair directly (map key -> value).
+`SC` (SMT) is used to commit to a data and compute merkle proofs. `SS` is used to directly access data. To avoid collisions, both `SS` and `SC` will use a separate storage namespace (they could use the same database underneath). `SS` will store each `(key, value)` pair directly (map key -> value).
 
-SMT is a merkle tree structure: we don't store keys directly. For every `(key, value)` pair, `hash(key)` is stored in a path (we hash a key to evenly distribute keys in the tree) and `hash(key, value)` in a leaf. Since we don't know a structure of a value (in particular if it contains the key) we hash both the key and the value in the `SC` leaf.
+SMT is a merkle tree structure: we don't store keys directly. For every `(key, value)` pair, `hash(key)` is stored in a path (we hash a key to evenly distribute keys in the tree) and `hash(value)` in a leaf.
```

Comment:
```suggestion
SMT is a merkle tree structure: we don't store keys directly. For every `(key, value)` pair, `hash(key)` is stored in a path (we hash a key to evenly distribute keys in the tree) and `hash(key) + hash(value)` in a leaf.
```

Reply:
as discussed - in the leaf we need to commit to the key as well (it's not enough that it is in a path).

Reply:
Path *is* stored in the leaf node, even if not as part of the "leaf value" (i.e. the value returned from calling `Get` on the SMT`

leaf node == `prefix || path || leaf_value` == `prefix || hash(key) || hash(value_provided_to_the_SMT)`

When calling `Set(key, value_provided_to_the_SMT)` it hashes the `key` into the `path` and the `value_provided_to_the_SMT` into the `leaf_value`. When calling `Get(key)` it returns `value_provided_to_the_SMT`.

Reply:
Right, so we can add `prefix +` to my suggestion. We can use `||` operator instead of `+` if you prefer.

> Note that value_provided_to_the_SMT currently is `hash(key || value_in_the_StateStore)`

Why is that? We should provide `key` and `obj_value` without modifying it. SMT will do all necessary operations.

Reply:
That was just a misunderstanding of the ADR language, I think. We didn't think it was describing the SMT's internal structure, since the hashed values are not exposed by the SMT interface. But we can add methods for that and fork the code if necessary.

Reply:
I think we are on the same page. In my suggestion I added `hash(key)` to the leaf (I'm using `+` operator rather than `||`). It seams we need to update it to (as you noted in the comment above):

    prefix + hash(key) + hash(value_provided_to_the_SMT)

Reply:
Ping, @roysc , @i-norden  - let's update the paragraph to what's in the SMT leaf and merge this PR.

Reply:
I have updated my suggestion based on the SMT spec: https://github.com/celestiaorg/celestia-specs/blob/ec98170398dfc6394423ee79b00b71038879e211/src/specs/data_structures.md#sparse-merkle-tree and John [response](https://github.com/cosmos/cosmos-sdk/discussions/9331#discussioncomment-996639).